Introduction to Free Monads/article.md

```diff
@@ -0,0 +1,702 @@
+Pure values, i.e. leaves, are encoded via the constructor `Pure`, and "actions", i.e. branches -- via the constructor `Free`.
+
+It may be helpful to write out the types of these constructors explicitly:
+
+```haskell
+Pure :: a -> Free f a
+Free :: f (Free f a) -> Free f a
+```
+
+If you squint a bit, you will notice that `Pure` has the same signature as `return` (assuming `Monad (Free f)`).
+If you squint some more, you will notice that the signature of `Free` is very similar to the signature of `join`:
+
+```haskell
+join :: Monad m => m (m a) -> m a
+```
+
+Instead of defining monads in terms of `return` and `>>=`, we can alternatively define them in terms of `map`, `return`, and `join`.
+So this hopefully provides some insight into why such a structure works as an encoding for a free monad.
+Indeed, if we require that `f` in `Free f` must be a functor, we get `fmap` from the start, and `Pure` and `Free` fill the roles of `return` and `join` respectively.
+
+If `f` is a `Functor`, then `Free f` is also a `Functor`.
+Indeed, the instance is very straightforward:
+
+```haskell
+instance Functor f => Functor (Free f) where
+  fmap f (Pure x) = Pure (f x)
+  fmap f (Free g) = Free (fmap f <$> g)
+```
+
+Notice that we need `f` to be a `Functor` to recursively descend the `Free` branch.
+
+If `f` is a `Functor`, then `Free f` is also a `Monad`.
+The instance is also pretty straightforward:
+
+```haskell
+instance Functor f => Monad (Free f) where
+  return = Pure
+  Pure x >>= f = f x
+  Free g >>= f = Free ((>>= f) <$> g)
+```
+
+Note that the implementation of `>>=` is very similar to that of `fmap`.
+If we instead defined the `Monad` instance in terms of `join`, it would look something like this:
+
+```haskell
+join (Pure x) = x
+join (Free x) = Free $ join <$> x
+```
+
+What must be noted here, is that this implementation, while straightforward, is not particularly efficient.
+A more efficient implementation, known as Church-encoded free monads, is available.
+Naturally, when using free monads, you should prefer the more efficient implementation.
+However, the core idea remains the same, so we're not delving into this topic in this introduction.
+
+Free monads, along with free applicative and alternative functors, and cofree comonads, are provided by the `free` package [on Hackage][hackage-link] (and [Stackage][stackage-link]).
+
+[hackage-link]: https://hackage.haskell.org/package/free
+[stackage-link]: https://www.stackage.org/lts/package/free
+
+Using free monads, we can define a computation as a data structure.
+Here, "computation" is a very broad term.
+The data structure in question doesn't define how the computation is performed, and we can write a separate interpreter (or indeed many interpreters) that performs the actual computation in question.
+
+## Using free monads
+
+> Talk is cheap, show me the code
```

Comment:
My gut tells me to cut this quote, it is rather overquoted & in comparison to others doesn't really bring anything new.

Reply:
I'll explain my thought process here. So while writing this, I've realized I just can't find it in me to not quote the "mostly wrong history of the programming languages", because I think it's brilliant, simultaneously funny and provides some deep insight into the underlying theory around monads. But then the only section with a quote stood out like a sore thumb, so I've decided to add relevant quotes to each section. The "free speech" vs "free beer" did come naturally right away. The other ones, not so much. So, for some definition of "relevant", this Linus quote is the most relevant to the "examples" section I could come up with (I also find it to be a little funny, so I feel it helps a bit with keeping the style more in the realm of "blog post" and less in the realm of "boring university lecture")

I do agree that the last two quotes don't bring much new to the table, but I also think dropping the quotes altogether halfway through would feel weird. I'm very open to suggestions for better quotes, however, so if you (or anyone else) has ideas, I'd very much like to hear those.

Reply:
I think that not every section needs to (necessarily) have a quote. So, if there is something that doesn't really add to the whole experience, it is ok to drop it & not strain oneself too much to find some quote, since (most of the) readers won't look at the article with the same kind of structural outlook. 

It is pretty much a judgement call though, and I won't be very unhappy if you decide to keep them.